client/normal_value.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package client
+
+import (
+	"time"
+
+	"github.com/sourcenetwork/immutable"
+)
+
+// NormalValue is the interface for the normal value types.
+// It is used to represent the normal (or standard) values across the system and to avoid
+// asserting all possible types like int, int32, int64, etc.
+//
+// All methods returning a specific type returns the value and the second boolean flag indicating
+// if the value is of the requested type.
+//
+// All nillable values are represented as immutable.Option[T].
+type NormalValue interface {
+	// Unwrap returns the underlying value.
+	// For not nillable values it will the value as is.
+	// For nillable values it will return result of `Value()` of `immutable.Option` if it
+	// has value, otherwise it will return `nil`.
+	Unwrap() any
+
+	// IsNil returns if the value is nil.
+	IsNil() bool
+	// IsNillable returns if the value can be nil.
+	IsNillable() bool
+
+	// Bool returns the value as a bool.
+	Bool() (bool, bool)
+	// Int returns the value as an int64.
+	Int() (int64, bool)
+	// Float returns the value as a float64.
+	Float() (float64, bool)
+	// String returns the value as a string.
+	String() (string, bool)
+	// Bytes returns the value as a byte slice.
+	Bytes() ([]byte, bool)
+	// Time returns the value as a time.Time.
+	Time() (time.Time, bool)
+	// Document returns the value as a *Document.
+	Document() (*Document, bool)
+
+	// NillableBool returns the value as a nillable bool.
+	NillableBool() (immutable.Option[bool], bool)
+	// NillableInt returns the value as a nillable int64.
+	NillableInt() (immutable.Option[int64], bool)
+	// NillableFloat returns the value as a nillable float64.
+	NillableFloat() (immutable.Option[float64], bool)
+	// NillableString returns the value as a nillable string.
+	NillableString() (immutable.Option[string], bool)
+	// NillableBytes returns the value as a nillable byte slice.
+	NillableBytes() (immutable.Option[[]byte], bool)
+	// NillableTime returns the value as a nillable time.Time.
+	NillableTime() (immutable.Option[time.Time], bool)
+	// NillableDocument returns the value as a nillable *Document.
+	NillableDocument() (immutable.Option[*Document], bool)
+
+	// IsArray returns if the value is an array.
+	IsArray() bool
```

Comment:
suggestion: `IsArray` might be slightly nicer to have futher up this definition, next to `IsNillable`.